File: csharp/DotnetLambdaBenchmark/src/DotnetLambdaBenchmark/FilePath.cs
using System;
using System.IO;

namespace DotnetLambdaBenchmark
{

    public class FilePath {

        public static FilePath FromNormalized(string normalizedPath) {
            var userID = normalizedPath.Substring(0, normalizedPath.IndexOf('/'));

            return new FilePath(userID, normalizedPath);
        }

        public static FilePath FromAbsolute(string absolutePath) {
            if (!Path.IsPathRooted(absolutePath)) {
                throw new ArgumentException("Invalid path, should be absolute");
            }

            // Normalizes the path and removes the leading slash
            var normalized = Path.GetFullPath(new Uri(absolutePath).LocalPath).TrimStart('/');
            var userID = normalized.Substring(0, normalized.IndexOf('/'));

            return new FilePath(userID, normalized);
        }

        private FilePath(string userID, string normPath) {
            if (normPath.Length <= userID.Length + 1) {
                throw new ArgumentException("Invalid path, missing name of the file");
            }

            this.UserID = userID;
            this.NormalizedPath = normPath;
        }

        public string NormalizedPath { get; }

        public string AbsolutePath => $"/{this.NormalizedPath}";

        public string Basename => Path.GetFileName(this.NormalizedPath);

        public string UserID { get; }
    }

}

File: java/src/main/java/benchmark/FilePath.java
package benchmark;

import java.nio.file.Path;
import java.nio.file.Paths;

class FilePath {

    public static FilePath fromAbsolute(String absolutePath) {
        Path path = Paths.get(absolutePath);
        if (!path.isAbsolute()) {
            throw new ArgumentException("Invalid path, should be absolute");
        }

        path = path.normalize();
        String userID = path.getName(0).toString();
        return new FilePath(userID, Path.of("/").relativize(path));
    }

    public static FilePath fromNormalized(String normalizedPath) {
        Path path = Paths.get(normalizedPath);
        String userID = path.getName(0).toString();
        return new FilePath(userID, path);
    }

    public String getNormalized() {
        return normalizedPath.toString();
    }

    public String getAbsolute() {
        return "/" + normalizedPath.toString();
    }

    public String getBasename() {
        return normalizedPath.getFileName().toString();
    }

    public String getUserID() {
        return userID;
    }

    private FilePath(String userID, Path normalizedPath) {
        if (normalizedPath.getNameCount() < 2) {
            throw new ArgumentException("Invalid path, missing name of the file");
        }

        this.userID = userID;
        this.normalizedPath = normalizedPath;
    }

    String userID;
    Path normalizedPath;
}
